# Grabbing data with cenpy

Cenpy (`sen - pie`) is a package that exposes APIs from the us Census Bureau and makes it easy to pull down and work with Census data in Pandas. First, notice that there are two core modules in the package, `base` and `explorer`, which each do different things. First, let's look at `explorer`. 

In [82]:
import cenpy.explorer as explorer
import pandas

On import, `explorer` request a current version of the APIs available from the Census Bureau. In future, it can also read a `JSON` collection describing the databases from disk, if needed.

Explorer has two functions, `available` and `explain`. `available` will provide a list of the identifiers of all the APIs that `cenpy` knows about. If run with `verbose=True`, `cenpy` will also include the title of the database as a dictionary. It's a good idea to *not* process this directly, and instead use it to explore currently available APIs. 

In [24]:
explorer.available(verbose=True)

{u'1990sf1': u'1990 Decennial: Summary File 1',
 u'2000sf1': u'2000 Decennial: Summary File 1',
 u'2000sf3': u'2000 Decennial: Summary File 3',
 u'2010acs5': u'2010 American Community Survey: 5-Year Estimates',
 u'2010sf1': u'2010 Decennial: Summary File 1',
 u'2011acs5': u'2011 American Community Survey: 5-Year Estimates',
 u'2012acs1': u'2012 American Community Survey: 1-Year Estimates',
 u'2012acs1profile': u'2012 American Community Survey: 1-Year Profile Tables',
 u'2012acs3': u'2012 American Community Survey: 3-Year Estimates',
 u'2012acs3profile': u'2012 American Community Survey: 3-Year Profile Tables',
 u'2012acs5profile': u'2012 American Community Survey: 5-Year Profile Tables',
 u'2012popproj/births': u'Vintage 2012 Population Projections - : Projected Births',
 u'ACSCD1132011': u'2011 American Community Survey 1-Year Profiles for the 113th Congressional Districts',
 u'ACSProfile2013': u'2013 American Community Survey - Data Profiles: 1-Year Data Profile',
 u'ACSProfile3Y2013

The `explain` command provides the title and full description of the datasource. If run in verbose mode, the function returns the full `json` listing of the API. 

In [25]:
explorer.explain('2010acs5')

(u'2010 American Community Survey: 5-Year Estimates',
 u"The American Community Survey (ACS) is a nationwide survey designed to provide communities a fresh look at how they are changing. The ACS replaced the decennial census long form in 2010 and thereafter by collecting long form type information throughout the decade rather than only once every 10 years.  Questionnaires are mailed to a sample of addresses to obtain information about households -- that is, about each person and the housing unit itself.  The American Community Survey produces demographic, social, housing and economic estimates in the form of 1-year, 3-year and 5-year estimates based on population thresholds. The strength of the ACS is in estimating population and housing characteristics. It produces estimates for small areas, including census tracts and population subgroups.  Although the ACS produces population, demographic and housing unit estimates,it is the Census Bureau's Population Estimates Program that produces

To actually connect to a database resource, you create a `Connection`. A `Connection` works like a *very* simplified connection from the `sqlalchemy` world. The `Connection` class has a method, `query` that constructs a query string and requests it from the Census server. This result is then parsed into JSON and returned to the user.  

In [83]:
import cenpy.base as base

In [26]:
conn = base.Connection('2010acs5')

In [27]:
conn

Connection to 2010 American Community Survey: 5-Year Estimates (ID: 2010acs5)

That may have taken longer than you'd've expected. This is because, when the `Connection` constructor is called, it populates the connection object with a bit of metadata that makes it possible to construct queries without referring to the census handbooks. 

For instance, a connection's `variables` represent all available search parameters for a given dataset. 

In [29]:
q.variables.head()

concept  \
AIANHH   Geographic Summary Level   
AIHHTLI                       NaN   
AITS                          NaN   
AITSCE                        NaN   
ANRC                          NaN   

                                                     label predicateOnly  \
AIANHH   American Indian Area/Alaska Native Area/Hawaii...           NaN   
AIHHTLI  American Indian Trust Land/Hawaiian Home Land ...           NaN   
AITS             American Indian Tribal Subdivision (FIPS)           NaN   
AITSCE         American Indian Tribal Subdivision (Census)           NaN   
ANRC             Alaska Native Regional Corporation (FIPS)           NaN   

        predicateType  
AIANHH            NaN  
AIHHTLI           NaN  
AITS              NaN  
AITSCE            NaN  
ANRC              NaN

This dataframe is populated just like the census's table describing the variables on the corresponding [api website](http://api.census.gov/data/2010/sf1/variables.html). Fortunately, this means that you can modify and filter this dataframe just like you can regular pandas dataframes, so working out what the exact codes to use in your query is easy. 

In [30]:
filt = [True if 'B000' in x else False for x in q.variables.index]

In [37]:
q.variables[filt]

concept  \
B00001_001E  B00001.  Unweighted Sample Count of the Popula...   
B00001_001M  B00001.  Unweighted Sample Count of the Popula...   
B00002_001E           B00002.  Unweighted Sample Housing Units   
B00002_001M           B00002.  Unweighted Sample Housing Units   

                                  label predicateOnly predicateType  
B00001_001E                       Total           NaN           NaN  
B00001_001M  Margin Of Error For!!Total           NaN           NaN  
B00002_001E                       Total           NaN           NaN  
B00002_001M  Margin Of Error For!!Total           NaN           NaN

Likewise, the different levels of geographic scale are determined from the metadata in the overall API listing and recorded.

In [11]:
q.geographies

geoLevelId                                               name  \
0         010                                                 us   
1         040                                              state   
2         050                                             county   
3         060                                 county subdivision   
4         067                         subminor civil subdivision   
5         070                                              place   
6         080                                              tract   
7         140                                              tract   
8         150                                        block group   
9         160                                              place   
10        250  american indian area/alaska native area/hawaii...   
11        280  american indian area/alaska native area/hawaii...   
12        500                             congressional district   

   optionalWithWCFor                                    requires  
0                NaN                                         NaN  
1                NaN                                         NaN  
2              state                                     [state]  
3                NaN                             [state, county]  
4                NaN         [state, county, county subdivision]  
5                NaN         [state, county, county subdivision]  
6                NaN  [state, county, county subdivision, place]  
7             county                             [state, county]  
8              tract                      [state, county, tract]  
9              state                                     [state]  
10               NaN                                         NaN  
11               NaN                                     [state]  
12             state                                     [state]

Note that some geographies have a **required** filter to prevent drawing too much data. This will get passed to the `query` method later. 

So, let's just grab the population and margins of error from the ACS. Using the variables table above, we picked out what fields we wanted. In addition, adding the `NAME` field smart-fills the table with the name of the geographic entity being pulled from the census bureau. 

In [38]:
cols = ['B00001_001E', 'B00001_001M', 'B00002_001E', 'B00002_001M', 'NAME'] 

Now the query. The query is constructed just like the API query, and works as follows. 

    cols - list of columns desired from the database
    geo_unit - string denoting the smallest geography to pull (unit of study)
    geo_filter - dictionary containing groupings of geo_units, if required
    
To be specific, a fully query tells the server *what* columns to pull of *what* underlying geography from *what* aggregation units. It's structured using these heterogeneous datatypes so it's easy to change the smallest units quickly, while providing sufficient granularity to change the filters and columns as you go. 

This query below grabs the names, population, and housing estimates from the ACS, as well as their standard errors from census designated places in Arizona. 

In [40]:
data = q.query(cols, geo_unit = 'place:*', geo_filter = {'state':'04'})

Once constructed, the query executes as fast as your internet connection will move. This query has:

In [41]:
data.shape

(451, 7)

Seven columns and 451 rows. So, rather fast. 

For validity and ease of use, we store the last executed query to the object. If you're dodgy about your census API key never being shown in plaintext, never print this property!

In [42]:
q.last_query

u'http://api.census.gov/data/2010/acs5?get=B00001_001E,B00001_001M,B00002_001E,B00002_001M,NAME&for=place:*&in=state:04'

Then, since the results are returned as a pandas dataframe, we can easily process the results. First, it's helpful to know that pandas will not automatically infer numbers from the information returned. So, if you want to convert a numeric column, first mask the data and then cast the column to an integer. 

Using this, I'll select all Census Designated Places in Arizona with populations above 5000.

In [43]:
data[data['B00001_001E'].fillna(0).apply(int) > 5000]

B00001_001E B00001_001M B00002_001E B00002_001M                      NAME  \
63        14249          -1        5698          -1    Chandler city, Arizona   
146       12256          -1        4335          -1     Gilbert town, Arizona   
148       13915          -1        5478          -1    Glendale city, Arizona   
224       28472          -1       12137          -1        Mesa city, Arizona   
266        9324          -1        3796          -1      Peoria city, Arizona   
268       84928          -1       34662          -1     Phoenix city, Arizona   
328       12973          -1        6929          -1  Scottsdale city, Arizona   
366        7253          -1        3042          -1    Surprise city, Arizona   
375        9311          -1        4151          -1       Tempe city, Arizona   
394       31499          -1       13967          -1      Tucson city, Arizona   
450        5841          -1        2215          -1        Yuma city, Arizona   

    state  place  
63     04  12000  
146    04  27400  
148    04  27820  
224    04  46000  
266    04  54050  
268    04  55000  
328    04  65000  
366    04  71510  
375    04  73000  
394    04  77000  
450    04  85540

## OK, that's one API, does it work for others?

In [44]:
conn2 = base.Connection('EconCensusEWKS2012')

In [48]:
filt2 = [True if 'Statistics' in x else False for x in conn2.variables['concept']]

In [69]:
conn2.variables[filt2]

concept  \
EMP              Employer Statistics   
EMP_F            Employer Statistics   
EMP_S            Employer Statistics   
EMP_S_F          Employer Statistics   
ESTAB            Employer Statistics   
ESTAB_F          Employer Statistics   
NESTAB       Non Employer Statistics   
NESTAB_F     Non Employer Statistics   
NRCPTOT      Non Employer Statistics   
NRCPTOT_F    Non Employer Statistics   
OPTAX            Employer Statistics   
PAYANN           Employer Statistics   
PAYANN_F         Employer Statistics   
PAYANN_S         Employer Statistics   
PAYANN_S_F       Employer Statistics   
PAYQTR1          Employer Statistics   
PAYQTR1_F        Employer Statistics   
PAYQTR1_S        Employer Statistics   
PAYQTR1_S_F      Employer Statistics   
RCPTOT           Employer Statistics   
RCPTOT_F         Employer Statistics   
RCPTOT_S         Employer Statistics   
RCPTOT_S_F       Employer Statistics   
YEAR             Employer Statistics   

                                                         label predicateOnly  \
EMP                                        Number of employees           NaN   
EMP_F        Flag for Number of employees data suppression ...           NaN   
EMP_S        Standard error for estimate of number of emplo...           NaN   
EMP_S_F      Flag for Standard error for estimate of number...           NaN   
ESTAB                                 Number of establishments           NaN   
ESTAB_F      Flag for Number of establishments data suppres...           NaN   
NESTAB                    Number of nonemployer establishments           NaN   
NESTAB_F     Flag for Number of nonemployer establishments ...           NaN   
NRCPTOT            Nonemployer value of business done ($1,000)           NaN   
NRCPTOT_F    Flag for Nonemployer value of business done ($...           NaN   
OPTAX                     Type of operation or tax status code           NaN   
PAYANN                                 Annual payroll ($1,000)           NaN   
PAYANN_F         Flag for Annual payroll data suppression flag           NaN   
PAYANN_S     Standard error for estimate of annual payroll (%)           NaN   
PAYANN_S_F   Flag for Standard error for estimate of annual...           NaN   
PAYQTR1                            Total first quarter payroll           NaN   
PAYQTR1_F    Flag for Total first quarter payroll data supp...           NaN   
PAYQTR1_S    Standard error for estimate of first-quarter p...           NaN   
PAYQTR1_S_F  Flag for Standard error for estimate of first-...           NaN   
RCPTOT                         Value of business done ($1,000)           NaN   
RCPTOT_F     Flag for Value of business done data suppressi...           NaN   
RCPTOT_S     Standard error for estimate of value of busine...           NaN   
RCPTOT_S_F   Flag for Standard error for estimate of value ...           NaN   
YEAR                                                      Year           NaN   

            predicateType required  
EMP                string      NaN  
EMP_F              string      NaN  
EMP_S              string      NaN  
EMP_S_F            string      NaN  
ESTAB              string      NaN  
ESTAB_F            string      NaN  
NESTAB             string      NaN  
NESTAB_F           string      NaN  
NRCPTOT            string      NaN  
NRCPTOT_F          string      NaN  
OPTAX              string     True  
PAYANN             string      NaN  
PAYANN_F           string      NaN  
PAYANN_S           string      NaN  
PAYANN_S_F         string      NaN  
PAYQTR1            string      NaN  
PAYQTR1_F          string      NaN  
PAYQTR1_S          string      NaN  
PAYQTR1_S_F        string      NaN  
RCPTOT             string      NaN  
RCPTOT_F           string      NaN  
RCPTOT_S           string      NaN  
RCPTOT_S_F         string      NaN  
YEAR                  int      NaN

In [47]:
conn2.geographies

geoLevelId                                               name  \
0        001                                                 us   
1        002                                              state   
2        007                                                csa   
3    008,009  metropolitan statistical area/micropolitan sta...   
4        004                                  consolidated city   
5        003                                             county   
6        006                                              place   
7        005                                              place   
8        010                              metropolitan division   

                                            requires  
0                                                NaN  
1                                                NaN  
2                                                NaN  
3                                                NaN  
4                                            [state]  
5                                            [state]  
6                                    [state, county]  
7                                            [state]  
8  [metropolitan statistical area/micropolitan st...

Here, pay special attention to the `required` column. Also, for this dataset, you need to provide additional parameters. To add however many additional paramters to your query, you add them as keyword arguments at the end. 

In [78]:
d2 = conn2.query(cols=['EMP', 'ESTAB'], geo_unit='place:*', geo_filter={'state':'06'}, OPTAX='*', NAICS2012 = '*')

In [79]:
conn2.last_query

u'http://api.census.gov/data/2012/ewks?get=EMP,ESTAB&NAICS2012=*&OPTAX=*&for=place:*&in=state:06'

In [80]:
d2

EMP ESTAB NAICS2012 OPTAX state  place
0        0     1        42    10    06  02812
1        0     1       424    10    06  02812
2        0     1      4249    10    06  02812
3        0     1     42493    10    06  02812
4        0     1    424930    10    06  02812
5        0     3     44-45    99    06  02812
6        0     1       445    99    06  02812
7        0     1      4451    99    06  02812
8        0     1     44511    99    06  02812
9        0     1    445110    99    06  02812
10       0     1       447    99    06  02812
11       0     1      4471    99    06  02812
12       0     1     44711    99    06  02812
13       0     1    447110    99    06  02812
14       0     1       454    99    06  02812
15       0     1      4541    99    06  02812
16       0     1     45411    99    06  02812
17       0     1     48-49    99    06  02812
18       0     1       484    99    06  02812
19       0     1      4842    99    06  02812
20       0     1     48422    99    06  02812
21       0     1    484220    99    06  02812
22       0     1   4842203    99    06  02812
23       0     3        56    99    06  02812
24       0     3       561    99    06  02812
25       0     1      5611    99    06  02812
26       0     1     56111    99    06  02812
27       0     1    561110    99    06  02812
28       0     1      5614    99    06  02812
29       0     1     56141    99    06  02812
...     ..   ...       ...   ...   ...    ...
212098   0     1     56161    99    06  98115
212099   0     1    561611    99    06  98115
212100   0     2      5617    99    06  98115
212101   0     1     56172    99    06  98115
212102   0     1    561720    99    06  98115
212103   0     1     56173    99    06  98115
212104   0     1    561730    99    06  98115
212105   0     1       562    99    06  98115
212106   0     1      5629    99    06  98115
212107   0     1     56291    99    06  98115
212108   0     1    562910    99    06  98115
212109   0    11        72    99    06  98115
212110   0     4       721    99    06  98115
212111   0     4      7212    99    06  98115
212112   0     4     72121    99    06  98115
212113   0     3    721211    99    06  98115
212114   0     1    721214    99    06  98115
212115   0     7       722    99    06  98115
212116   0     2      7223    99    06  98115
212117   0     1     72231    99    06  98115
212118   0     1    722310    99    06  98115
212119   0     1     72233    99    06  98115
212120   0     1    722330    99    06  98115
212121   0     1      7224    99    06  98115
212122   0     1     72241    99    06  98115
212123   0     1    722410    99    06  98115
212124  68     4      7225    99    06  98115
212125  68     4     72251    99    06  98115
212126   0     2    722511    99    06  98115
212127   0     2    722513    99    06  98115

[212128 rows x 6 columns]

And so you get the table of NAICS data for employment and establishments in places in California. 

## ToDo:

1. Provide meaningful feedback from the census errors in the reporting
2. Provide joins with the geodata from the TIGER rest api
3. Provide more convenience functions, like lookups for place and state abbreviations to fips codes?